cpp/src/Ice/ConnectionI.cpp

```diff
@@ -1511,33 +1526,44 @@ Ice::ConnectionI::message(ThreadPoolCurrent& current)
                 break;
             }
 
+            // readOp and writeOp are set to the operations that the transport read or write calls from above returned.
+            // They indicate which operations will need to be monitored by the thread pool's selector when this method
+            // returns.
             SocketOperation newOp = static_cast<SocketOperation>(readOp | writeOp);
-            readyOp = static_cast<SocketOperation>(readyOp & ~newOp);
+
+            // Operations that are ready. For example, if message was called with SocketOperationRead and the transport
+            // read returned SocketOperationNone, reads are considered done: there's no additional data to read.
+            SocketOperation readyOp = static_cast<SocketOperation>(current.operation & ~newOp);
             assert(readyOp || newOp);
 
             if (_state <= StateNotValidated)
             {
+                // If the connection is still not validated and there's still data to read or write, continue waiting
+                // for data to read or write.
                 if (newOp)
                 {
                     //
-                    // Wait for all the transceiver conditions to be
-                    // satisfied before continuing.
+                    // Wait for all the transceiver conditions to be satisfied before continuing.
                     //
                     scheduleTimeout(newOp);
                     _threadPool->update(shared_from_this(), current.operation, newOp);
                     return;
                 }
 
+                // Initialize the connection if it's not initialized yet.
                 if (_state == StateNotInitialized && !initialize(current.operation))
                 {
                     return;
                 }
 
+                // Validate the connection if it's not validate yet.
                 if (_state <= StateNotValidated && !validate(current.operation))
                 {
                     return;
                 }
 
+                // The connection is validated and doesn't need additional data to be read or written. So unregister
```

Comment:
It's not clear to me what this means: is this the expected condition once the connection is validated?
But then, how do we receive the first incoming request on this connection?

Reply:
The connection is in the holding state after being validated so we unregister the connection from the threadpool's selector. The connection will be activated by the connection factory (see activate calls on connections in `ConnectionFactory.cpp`).

Reply:
Would be nice to insert this clarification in the new comment.

So when you create an OA, it accepts connections _before_ you activate this OA?
Whereas with IceRPC, a server only accepts connections after you call Listen:
https://docs.icerpc.dev/api/csharp/api/IceRpc.Server.html#IceRpc_Server_Listen

How does this work for new connections on an activated OA? Somehow after this line the OA actives the connection and re-register it with the thread pool?

And how does it work for outgoing connections?

Reply:
> Would be nice to insert this clarification in the new comment.

Fixed.

> So when you create an OA, it accepts connections before you activate this OA?

When you create the OA, an incoming connection factory is created for each endpoint. An incoming connection factory creates the transport acceptor and the acceptor starts listening for connections. So the OA does indeed starts listening for connections on creation but it doesn't accept them just yet.  Connections are only accepted once the `ObjectAdapter::activate` method is called. The OA activate ends up registering each endpoint server socket with the threadpool's selector.

> How does this work for new connections on an activated OA? Somehow after this line the OA actives the connection and re-register it with the thread pool?

Connections are accepted by `IncomingConnectionFactory::message`: https://github.com/zeroc-ice/ice/blob/73af7f1cab4d6e4b73bdc8f1dc04f09fa18e6e7b/cpp/src/Ice/ConnectionFactory.cpp#L1405 This method is the same as for a connection, it's called by the thread pool when the TCP server socket is ready for accepting a connection.

This method creates the connection with the accepted transport transceiver. The connection is "started" which means it's initialized and then validated. Once done, the `IncomingConnectionFactory::connectionStartCompleted` is called:  https://github.com/zeroc-ice/ice/blob/73af7f1cab4d6e4b73bdc8f1dc04f09fa18e6e7b/cpp/src/Ice/ConnectionFactory.cpp#L1612

The connection is activated there only if the OA is not in the hold state.

> And how does it work for outgoing connections?

It's always activated after the connection start (initialize + validate) completes: https://github.com/zeroc-ice/ice/blob/73af7f1cab4d6e4b73bdc8f1dc04f09fa18e6e7b/cpp/src/Ice/ConnectionFactory.cpp#L891